### ISSUE TITLE
how to multinomial?

### ISSUE THREAD

**Issue by b-rodrigues** [greta-dev/greta#120]
Hi,

first of all, thanks for `greta`, this package looks really awesome! Can't wait for my new computer with a Nvidia 1080GTX to arrive to run models on it :)

My question is the following: I'm trying to estimate a multinomial logit model using the `Catsup` data from the `mlogit` package. Is there a function such as `ilogit()` for multinomial logit models or should I write my own function to return the probabilities for each alternative? Something such as:

```
value = beta0*ones(2798, 1) + 
  beta1*disp.heinz41 + 
  beta2*feat.heinz41 +
  beta3*price.heinz41 +
  beta4*disp.heinz32 +
  beta5*feat.heinz32 +
  beta6*price.heinz32 +
  beta7*disp.heinz28 + 
  beta8*feat.heinz28 +
  beta9*price.heinz28 +
  beta10*disp.hunts32 +
  beta11*feat.hunts32 +
  beta12*price.hunts32

prob = return_probs(value)
  
distribution(choice) = categorical(prob, dim = 1)
```

Where `return_probs()` is a user written function that returns the probabilities of each alternative for individual `i`?

I guess the data should be in long format instead of wide, right? Meaning one column for price, one for feat and one for disp with different levels, instead of disp.heinz41, disp.heinz32 etc....

Thanks in advance!

**Comment by goldingn** [comment]
Hi Bruno,

Awesome, I look forward to hearing how you get on with the new GPU!

I'm not entirely sure I understand (so please correct me if I'm on the wrong track) but it sounds like you want something like the [softmax function](https://en.wikipedia.org/wiki/Softmax_function).

That's not implemented yet, but certainly could and should be! As a temporary fix, I've defined a version below that you can use. I've also opened an issue (#124 ) to add this to greta.

```
library (greta)
library(tensorflow)
op <- greta::.internals$nodes$constructors$op
softmax <- function (logits) {
  op("softmax", logits, tf_operation = tf$nn$softmax)
}
```

Unfortunately, the current implementation of the `categorical()` only accepts a *vector* of parameters for `prob`; it assumes that all rows on the left hand side follow the distribution with the same vector of probabilities. I've opened an issue (#123) to resolve that one too. When that's implemented that you should be able to define a categorical model like this:

```
n <- 10  # observations
k <- 5 # choices
m <- 3 # covariates

# fake data
covariates <- matrix(rnorm(n * m), n, m)
choices <- t(rmultinom(n, 1, rep(0.5, k)))

# create design matrix
X <- cbind(1, covariates)

# regression coefficients, one column for each choice 
beta <- normal(0, 1, dim = c(m + 1, k))

# multiply to get n-by-k matrix of unscaled logit-probabilities
logits <- X %*% beta

# convert to n-by-k matrix of probabilities (rows sum to 1)
prob <- softmax(logits)

# define the likelihood (this bit doesn't work yet, because of issue #123)
distribution(choice) = categorical(prob, dim = n)
```
I'll leave this issue open until I've implemented those both and this is working for you

**Comment by goldingn** [comment]
FYI, you could fit the above model by doing this instead of the final line:
```
for (i in seq_len(n)) {
  y <- t(choices[i, ])
  distribution(y) = categorical(t(prob[i, ]))
}
```
but it will be very slow if you have lots of observations. The matrix version will be much faster!

**Comment by b-rodrigues** [comment]
Great thank you very much! I'll test the dev version and give feedback :)

**Comment by goldingn** [comment]
I just released version 0.2.4 of greta on GitHub (`devtools::install_github("greta-dev/greta")`) which contains a `imultilogit()` transformation function which is similar to the softmax function, and designed for this purpose. Unlike softmax, `imultilogit()` takes as input a matrix with **n-1** columns, where there are n choices.
That enables a fully-identified linear model for multinomial data (unlike softmax).

The matrix-parameter multinomial function is still on the to do list!

**Comment by goldingn** [comment]
As detailed [here](https://github.com/greta-dev/greta/issues/123#issuecomment-413072893) there's a version of matrix-parametered multivariate distributions implemented and ready for testing.

Here's a working example, like the one above, but using that branch:

```r
library(greta)

n <- 10  # observations
k <- 5 # choices
m <- 3 # covariates

# fake data
covariates <- matrix(rnorm(n * m), n, m)
choices <- t(rmultinom(n, 1, rep(0.5, k)))

# create design matrix
X <- cbind(1, covariates)

# regression coefficients, one row for each covariate (including the intercept),
# one column for each choice, except the reference choice
beta <- normal(0, 1, dim = c(m + 1, k - 1))

# multiply to get n-by-(k - 1) matrix of unscaled logit-probabilities
logits <- X %*% beta

# convert to n-by-k matrix of probabilities (rows sum to 1)
prob <- imultilogit(logits)

# define the likelihood
distribution(choices) <- categorical(prob, n_realisations = n)

m <- model(beta)
draws <- mcmc(m, warmup = 2000)
plot(draws)
```

**Comment by goldingn** [comment]
I think this is fixed now, so I'm closing this and a bunch of other issues
